source/client-side-encryption/client-side-encryption.rst

```diff
@@ -770,73 +942,299 @@ behind-the-scenes operations required for encryption or decryption.
+``mongo_csfle_v1`` (with the appropriate file extension or filename suffix for
+the host platform). This library will be loaded by libmongocrypt when the
+``mongocrypt_init`` function is invoked
+`(from the libmongocrypt C API) <lmc-c-api_>`_ based on the search critera that
+are provided by the driver.
+
+libmongocrypt allows the driver to specify an arbitrary list of directory paths
+in which to search for the `csfle` dynamic library. The user may specify `csfle`
+search behavior by specifying options in `cse.extraOptions`.
+
+.. note::
+
+   The driver MUST NOT manipulate or do any validation on the csfle path options
+   provided in `cse.extraoptions`. They should be passed through to
+   libmongocrypt unchanged.
+
+
+Path Resolution Behavior
+------------------------
+
+.. |---| unicode:: U+2014
+
+Drivers should include and note the following information regarding the behavior
+of `csfle` path options in `cse.extraOptions`:
+
+- If used, the `override path`_ must be given as a path to the `csfle` dynamic
+  library file *itself*, and not simply the directory that contains it.
+
+- If the `override path`_ is specified, the `search paths`_ will be ignored.
+
+- For `csfle` `search paths`_ and the `override path`_, if the given path is a
+  relative path and the first path component is the literal string
+  "``$ORIGIN``", the "``$ORIGIN``" component will be replaced by the absolute
+  path to the directory containing the `libmongocrypt` library that is
+  performing the `csfle` search. This behavior mimics the ``$ORIGIN`` behavior
+  of the ``RUNPATH``/``RPATH`` properties of ELF executable files. This permits
+  bundling the `csfle` library along with `libmongocrypt` for creating portable
+  application distributions without relying on a externally/globally available
+  `csfle` library.
+
+  .. note:: No other ``RPATH``/``RUNPATH``-style substitutions are available.
+
+  .. _cse.csfle.dollar-system:
+
+- For `csfle` `search paths`_, if a search path string is "``$SYSTEM``", then
+  |---| instead of `libmongocrypt` searching for `csfle` in a directory named
+  "``SYSTEM``" |---| `libmongocrypt` will defer to the operating system's own
+  dynamic-library resolution mechanism when processing that search-path.
+
+- If a `search path`_ or `override path`_ is given as a relative path, that path
+  will be resolved relative to the working directory of the operating system
+  process.
+
+.. _search path:
+.. _search paths:
+.. _cse.csfle.search-paths:
+
+Adding Search Paths
+-------------------
+
+The driver MUST append csfle search paths to the `libmongocrypt_handle` in the
+following order:
+
+1. Append each search path in |opt-paths-prefer|.
+2. If `extraOptions.csfleDisableSystemLibrary <cse.extraoptions_>`__ *was not*
+   specified as ``true`` by the user, append the literal string "``$SYSTEM``" to
+   the path list. (`Refer <cse.csfle.dollar-system_>`_)
+3. Append each search path in |opt-paths-fallback|.
```

Comment:
In the server shared library planning docs, the expectation was that, in typical situations, the shared library would be deployed as part of application directories, not as part of the system libraries. 

Considering this, would it be worth simplifying this interface to merge the csfleSearchPathsPrefer/csfleSearchPathsFallback/csfleDisableSystemLibrary options and only specify a single list of search paths, which may or may not include `$SYSTEM`? We already need to make users aware that `$ORIGIN` exists as an expansion.

(If the desire here is to make the shared library an opt-out feature rather than an opt-in one, we could make specifying an empty search path list different from specifying no search path list at all, i.e. make the default `["$SYSTEM"]` rather than `[]`.)

Reply:
I interpretted the discussion as to whether it would be system-installed to be still tentative (although I would fall on the side of application-level deployment rather than system installation).

Given that we already need to document search path behavior, I think it would be feasible to merge these options and allow "set an empty path list" to be the "disable csfle" option. It will minimize the moving parts.

Although that still leaves an open question as to what the default should be (as related to your other comment). A default of `["$SYSTEM"]` is reasonable if we prefer a system-wide installation, but a default of `["$ORIGIN"]` would be better for application-level deployment (i.e. the user would only need to have `csfle` reside alongside the `libmongocrypt` library).

Either way, I like the idea to reduce the moving parts and merge these options and replace the "disable" toggle.